File: example/1d-linear-convection/weno3/python/04k/domain.py
# domain.py
from mesh import Mesh

class Domain:
    """计算域：管理物理区域、ghost层、索引映射等逻辑，依赖 Mesh 提供几何信息"""
    def __init__(self, config, mesh):
        """
        初始化计算域
        :param mesh: Mesh实例（静态网格属性）
        :param config: CfdConfig实例（包含recon_scheme/spatial_order）
        """
        self.config = config
        self.mesh = mesh        
        
        # 核心：根据重建格式动态计算nghosts
        self.nghosts = self._calc_nghosts()
        
        # 基于nghosts推导索引
        self.ist = self.nghosts  # 物理网格起始索引
        self.ied = self.ist + mesh.ncells  # 物理网格结束索引
        self.ntcells = mesh.ncells + 2 * self.nghosts  # 总网格数（含ghost）
       
        # 可选：调试信息（可后续移除）
        # print(f"mesh.ncells={mesh.ncells}")
        # print(f"self.config.spatial_order={self.config.spatial_order}")
        # print(f"self.nghosts={self.nghosts}")
        # print(f"self.ist={self.ist}")
        # print(f"self.ied={self.ied}")
        
    def _calc_nghosts(self):
        """内部方法：根据重建格式和阶数计算ghost层数量"""
        scheme = self.config.recon_scheme
        order = self.config.spatial_order
        
        if scheme is None:
            raise ValueError("必须先通过 with_reconstruction 设置重建格式！")
        
        if scheme == "eno":
            nghosts = order
        elif scheme == "weno":
            nghosts = order // 2 + 1
        else:
            raise ValueError(f"未知重建格式 {scheme}，无法计算ghost层！")
        
        if nghosts <= 0:
            raise ValueError(f"计算得到的ghost层数量无效：{nghosts}（阶数{order}，格式{scheme}）")
        return nghosts        
        
    def is_physical_cell(self, idx):
        """判断索引是否在物理网格范围内"""
        return self.ist <= idx < self.ied
    
    def get_physical_indices(self):
        """返回物理网格的索引范围（可直接用于循环）"""
        return range(self.ist, self.ied)

File: example/1d-linear-convection/weno3/fortran/registry/03g/src/infrastructure/domain.f90
! src/infrastructure/domain.f90（修正版）
module domain_module
    use base_modules, only: wp, ip, max_name_len
    use config_module, only: cfd_config
    use mesh_module, only: mesh_type
    
    implicit none
    private
    public :: wp, ip, domain_type, domain_create, is_physical_cell
    
    type :: domain_type
        type(cfd_config), pointer :: config => null()
        type(mesh_type), pointer :: mesh => null()
        integer(ip) :: nghosts = 0
        integer(ip) :: ist = 1      ! 物理区域起始索引（1-based）
        integer(ip) :: ied = 1      ! 物理区域结束索引（exclusive）
        integer(ip) :: ntcells = 0  ! 总单元数（含ghost）
    contains
        procedure :: print_info => domain_print_info
        procedure :: get_physical_indices => domain_get_physical_indices
    end type domain_type
    
contains

    function domain_create(config, mesh) result(domain)
        type(cfd_config), target, intent(in) :: config
        type(mesh_type), target, intent(in) :: mesh
        type(domain_type) :: domain
        
        domain%config => config
        domain%mesh => mesh
        
        ! 计算ghost层数（参考Julia的_calc_nghosts）
        domain%nghosts = calc_nghosts(config)
        domain%ist = domain%nghosts + 1
        domain%ied = domain%ist + mesh%ncells
        domain%ntcells = mesh%ncells + 2 * domain%nghosts
        
        if (config%verbose) then
            print *, "[DOMAIN] Created:"
            print *, "  Ghost layers: ", domain%nghosts
            print *, "  Physical cells: ", domain%ist, " to ", domain%ied - 1
            print *, "  Total cells: ", domain%ntcells
        end if
    end function domain_create
    
    function calc_nghosts(config) result(nghosts)
        type(cfd_config), intent(in) :: config
        integer(ip) :: nghosts
        
        character(len=max_name_len) :: scheme
        
        scheme = config%recon_scheme
        
        if (scheme == "eno") then
            nghosts = config%spatial_order
        else if (index(scheme, "weno") > 0) then
            nghosts = config%spatial_order / 2 + 1
        else
            print *, "[WARNING] Unknown scheme, using default nghosts=2"
            nghosts = 2
        end if
        
        if (nghosts <= 0) then
            print *, "[ERROR] Invalid nghosts: ", nghosts
            nghosts = 2
        end if
    end function calc_nghosts
    
    logical function is_physical_cell(this, idx)
        class(domain_type), intent(in) :: this
        integer(ip), intent(in) :: idx
        is_physical_cell = (idx >= this%ist .and. idx < this%ied)
    end function is_physical_cell
    
    function domain_get_physical_indices(this) result(indices)
        class(domain_type), intent(in) :: this
        integer(ip), allocatable :: indices(:)
        integer(ip) :: i, count
        
        count = this%ied - this%ist
        allocate(indices(count))
        
        do i = 1, count
            indices(i) = this%ist + i - 1
        end do
    end function domain_get_physical_indices
    
    subroutine domain_print_info(this)
        class(domain_type), intent(in) :: this
        
        print *, "=== Domain Information ==="
        print *, "Configuration: ", trim(this%config%recon_scheme), &
                 " order ", this%config%spatial_order
        print *, "Ghost layers: ", this%nghosts
        print *, "Physical cells: ", this%ist, " to ", this%ied - 1
        print *, "Total cells: ", this%ntcells
        print *, "Mesh cells: ", this%mesh%ncells
        print *, "=========================="
    end subroutine domain_print_info

end module domain_module